crates/bitwarden-threading/src/cancellation_token.rs

```diff
@@ -0,0 +1,66 @@
+pub use tokio_util::sync::CancellationToken;
+
+#[cfg(target_arch = "wasm32")]
+pub mod wasm {
+    use wasm_bindgen::prelude::*;
+
+    use super::*;
+
+    #[wasm_bindgen]
+    extern "C" {
+        #[wasm_bindgen]
+        #[derive(Clone)]
+        pub type AbortController;
+
+        #[wasm_bindgen(constructor)]
+        pub fn new() -> AbortController;
+
+        #[wasm_bindgen(method, getter)]
+        pub fn signal(this: &AbortController) -> AbortSignal;
+
+        #[wasm_bindgen(method, js_name = abort)]
+        pub fn abort(this: &AbortController, reason: JsValue);
+
+        #[wasm_bindgen]
+        pub type AbortSignal;
+
+        #[wasm_bindgen(method, getter)]
+        pub fn aborted(this: &AbortSignal) -> bool;
+
+        #[wasm_bindgen(method, js_name = addEventListener)]
+        pub fn add_event_listener(
+            this: &AbortSignal,
+            event_type: &str,
+            callback: &Closure<dyn FnMut()>,
+        );
+    }
+
+    pub trait AbortControllerExt {
+        fn to_cancellation_token(&self) -> CancellationToken;
+    }
+
+    impl AbortControllerExt for AbortController {
+        fn to_cancellation_token(&self) -> CancellationToken {
+            self.signal().to_cancellation_token()
+        }
+    }
+
+    pub trait AbortSignalExt {
+        fn to_cancellation_token(&self) -> CancellationToken;
+    }
+
+    impl AbortSignalExt for AbortSignal {
+        fn to_cancellation_token(&self) -> CancellationToken {
+            let token = CancellationToken::new();
+
+            let token_clone = token.clone();
+            let closure = Closure::new(move || {
+                token_clone.cancel();
+            });
+            self.add_event_listener("abort", &closure);
+            closure.forget(); // Transfer ownership to the JS runtime
+
+            token
+        }
+    }
```

Comment:
We should probably add a basic test to ensure this works:
```rust
        fn should_abort_wasm() {
            use super::*;

            console_error_panic_hook::set_once();

            let abort = AbortController::new();
            let signal = abort.signal();

            let token = signal.to_cancellation_token();

            assert!(!token.is_cancelled(), "Token should not be cancelled");
            assert!(!signal.aborted(), "Signal should not be aborted");

            abort.abort(JsValue::from("Test reason"));

            assert!(token.is_cancelled(), "Token should be cancelled");
            assert!(signal.aborted(), "Signal should be aborted");
        }
```

Reply:
Also, I wonder if we should add a comment indicating that the returned CancellationToken is only for receiving the cancellation signal. Currently you can do something like this:
```rust
let abort = AbortController::new();
let signal = abort.signal();

// This will do nothing, you have to use abort.abort();
signal.to_cancellation_token().cancel();

// Still not aborted
assert!(!signal.aborted(), "Signal is not aborted");
```

Reply:
You mean this? Or is there a subtle difference I'm missing?

https://github.com/bitwarden/sdk-internal/blob/b7115f6f96994d50f648aaab4998080f956efe8c/crates/bitwarden-threading/tests/cancellation_token/wasm.rs#L33-L51

Added comment!

Reply:
Hehe, decided to put them there so I can test the public interface. I've forgotten `pub` too many times